### PR TITLE
Ratio selector improvements

### PR DESCRIPTION
**Pull request by sallar**
Still to do:

+ [x] Custom Ratio should update when crop area is resized manually
+ [x] Sometimes Custom Ratio shows weird numbers
+ [x] Apps list should update when they come into view (or Kap window is shown again)
+ [x] When you select an app in the ratio dropdown, I think it should show the app name and icon in the unopened dropdown view instead of Custom (720:239). Same when Fullscreen is selected.
+ [x] Select "16:9" ratio, then write 100 in the first input. Then press the switch button, to switch the ratio. Now it says Custom (14:25), while it should have said Custom (9:16). This is working correctly in Kap Beta.
+ [x] The ratio dropdown should always open in the same place, like previously, and not where the mouse is.

image

image

image

### CODE REVIEW

**Comment by sindresorhus** [comment]
This is really great!

**Comment by sindresorhus** [comment]
- The ratio dropdown should always open in the same place, like previously, and not where the mouse is.
- When you select an app in the ratio dropdown, I think it should show the app name and icon in the unopened dropdown view instead of `Custom (720:239)`. Same when `Fullscreen` is selected.

File: app/src/renderer/js/size-selector.js [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+function updateContent(el, knownRatio, currentRatio) {
+  currentRatio = currentRatio.join(':');
+  el.querySelector('button').innerHTML = knownRatio || `Custom (${currentRatio})`;
+}
+
+function handleAppChange(app) {
+  if (app.fullscreen) {
+    ipcRenderer.send('open-cropper-window', {width: app.width, height: app.height}, {x: 1, y: 1});
+  } else {
+    ipcRenderer.send('activate-application', app.ownerName, app);
```
Comment (sindresorhus) [inline app/src/renderer/js/size-selector.js]: `activate-application` => `activate-app`

**Comment by sindresorhus** [comment]
- Select "16:9" ratio, then write `100` in the first input. Then press the switch button, to switch the ratio. Now it says `Custom (14:25)`, while it should have said `Custom (9:16)`. This is working correctly in Kap Beta.

**Comment by sallar** [comment]
Thansk @sindresorhus I added your comments to the checklist above. About opening the menu in the same place like before, I have to give it the coordinates of the "selector el" on the page. so it opens at the same place every time. but there is the issue of the menu not having the same width as the el...

File: app/src/renderer/js/size-selector.js [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+  const images = await getAppIconListByPid(windows.map(win => win.pid), {
+    size: 16,
+    encoding: 'buffer'
+  });
+  const {width, height} = remote.screen.getPrimaryDisplay().bounds;
+
+  return [
+    {
+      ownerName: 'Fullscreen',
+      pid: -10,
```
Comment (sindresorhus) [inline app/src/renderer/js/size-selector.js]: Don't abuse `pid` for something unrelated.

**Comment by sindresorhus** [comment]
- The dropdown menu should be attached to the top-right of the dropdown widget, not the left side. Try some native dropdowns and see.
- If I select an app window and then select a ratio, the cropper doesn't change to the new ratio if I have the "lock ratio" button selected. If "lock ratio" is unchecked, it the cropper updates, but the height of it is incorrect.

File: app/src/renderer/js/main.js [hunk 3]
```diff
@@ -241,50 +176,21 @@ document.addEventListener('DOMContentLoaded', () => {
...
-      customRatio.textContent = `Custom (${stringRatio})`;
-      customRatio.selected = true;
+    // The width and height inputs are not producing whole numbers
+    // sometimes they put out strings, sometimes floats and sometimes ints.
+    // parseInt is used here so the ratio calculation doesn't die.
+    width = parseInt(width, 10);
+    height = parseInt(height, 10);
+    dimensions.ratio = findRatioForSize(width, height);
+
+    // Remove pid from dimensions object
```
Comment (sindresorhus) [inline app/src/renderer/js/main.js]: I think you mean `app`, not `pid`.

File: app/src/renderer/js/size-selector.js [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    const app = windowList.find(win => win.pid === dimensions.app.pid);
+    if (app) {
+      content.innerHTML = getAppDisplayName(app);
+      return;
+    }
+  }
+
+  const stringRatio = dimensions.ratio.join(':');
+  const knownRatio = RATIOS.find(ratio => ratio === stringRatio);
+  content.innerHTML = knownRatio || `Custom (${stringRatio})`;
```
Comment (sindresorhus) [inline app/src/renderer/js/size-selector.js]: `innerHTML` => `textContent`

File: app/src/renderer/js/size-selector.js [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+      enabled: false,
+      type: 'checkbox',
+      checked: !knownRatio && !currentDimensions.app
+    }
+  ]);
+}
+
+function sizeMatchesRatio(width, height, ratio) {
+  const [first, second] = ratio.split(':');
+  return (width / first === height / second);
```
Comment (sindresorhus) [inline app/src/renderer/js/size-selector.js]: Remove the moot parens.

**Comment by sindresorhus** [comment]
> If I select an app window and then select a ratio, the cropper doesn't change to the new ratio if I have the "lock ratio" button selected. If "lock ratio" is unchecked, it the cropper updates, but the height of it is incorrect.

It now updates when "lock ratio" is checked, but the height is incorrect. If I choose "System Preferences" as an app window, which is 600x600 and then update the width to 601, the height in the cropper suddenly changes to 100.

**Comment by sindresorhus** [comment]
Here's how it looks like then:

screen shot 2018-01-07 at 17 37 05

File: app/src/renderer/js/main.js [hunk 6]
```diff
@@ -406,25 +303,48 @@ document.addEventListener('DOMContentLoaded', () => {
...
+  ipcRenderer.on('change-aspect-ratio', (e, aspectRatio) => handleSizeChange(aspectRatio));
+
+  dimensionsEmitter.on('ratio-selected', ratio => handleSizeChange(ratio));
+
+  dimensionsEmitter.on('app-selected', app => {
+    // Set app information on dimensions so it can be reused later
+    // eg. for showing app name after selection
+    dimensions.app = {
+      pid: app.pid,
+      fullscreen: app.fullscreen,
```
Comment (sindresorhus) [inline app/src/renderer/js/main.js]: Super nitpick, but `fullscreen` => `isFullscreen` (To make it clear when reading the code that it's a boolean).

**Comment by sallar** [comment]
@sindresorhus I couldn't reproduce

image

what is the correct behavior?

**Comment by sindresorhus** [comment]
> what is the correct behavior?

The width and height of the cropper should increase by one pixel, but instead, the height totally changed.

**Comment by sallar** [comment]
@sindresorhus Aha! Figured out. Please check now :)

**Comment by sindresorhus** [comment]
Great, that fixed it :)

Now there's another problem. If I first select an app window in the dropdown, and then select `16:9` ratio, it doesn't show `16:9` in the dropdown widget:

screen shot 2018-01-08 at 10 01 03

**Comment by sallar** [comment]
@sindresorhus thats because of the number rounding. I couldnt fix that... Because 16:9 needs floating points to work with some widths and heights :( Fixing that is whole other beast.

**Comment by sindresorhus** [comment]
Alright. That should block this. I've opened some new issues about the ratio problems: 

- https://github.com/wulkano/kap/issues/316
- https://github.com/wulkano/kap/issues/315
- https://github.com/wulkano/kap/issues/310

**Comment by sallar** [comment]
@sindresorhus bear in mind that this PR includes a lot of fixes to the ratios etc. so those issues need to come "after" this PR.

**Comment by sindresorhus** [comment]
@sallar Yes, I agree, that's what I said.

**Comment by sallar** [comment]
@sindresorhus I misunderstood then. great.

**Comment by sindresorhus** [comment]
Ok, just noticed I wrote `should` instead of `shouldn't`. Ugh.

**Comment by sindresorhus** [comment]
Heroic effort on this PR, @sallar ✨

superhighfive

**Pull request merged**